staging/src/k8s.io/apiserver/pkg/util/flowcontrol/rule.go

```diff
@@ -153,7 +153,19 @@ func matchPolicyRuleVerb(policyRuleVerbs []string, requestVerb string) bool {
 }
 
 func matchPolicyRuleNonResourceURL(policyRuleRequestURLs []string, requestPath string) bool {
-	return containsString(requestPath, policyRuleRequestURLs, fctypesv1a1.NonResourceAll)
+	for _, rulePath := range policyRuleRequestURLs {
+		if rulePath == fctypesv1a1.NonResourceAll || rulePath == requestPath {
+			return true
+		}
+		rulePrefix := strings.TrimSuffix(rulePath, "*")
```

Comment:
We need a validation change too, no? Otherwise "/foo/bar*" will be acted upon as if it were "/foo/bar/*".

Side note: string manipulation like this produces a lot of garbage, when we start caring about performance more we'll want to pre-compute this.

Reply:
The validation is already there.
I agree, there are performance improvements to be made in several ways.

Reply:
the url validation: https://github.com/kubernetes/kubernetes/blob/edf460bc0f50822e3cb123457445aa63ae87cf10/pkg/apis/flowcontrol/validation/validation.go#L470-L497

in terms of string garbages, we can simplify the matching logic given that the validation ensures that wildcard can only present in the end.

```
func nonResourceURLRuleWildCardMatch(rule, target string) bool {
	for i := 0; i < len(rule); i++ {
		if rule[i] == '*' {
			return true
		}
		if i >= len(target) {
			return false
		}
		if rule[i] != target[i] {
			return false
		}
	}
	return true
}
```